lib/veri.ml

```diff
@@ -148,41 +152,33 @@ class ['a] t arch dis =
     method private update_event ev =
       SM.update (fun c -> c#register_event ev)
 
-    (** [resolve_var var] - returns a result, bound with [var].
-        Sequence of searches is the following:
-        1) among read events that occured at current step in the same context,
-           with the same variable;
-        2) among read events that occures at current step, in other context,
-           with the same variable;
-        3) in current context, for the same variable *)
-    method private resolve_var : var -> 'a r = fun var ->
+  (** [find_var_data var] - returns a result, bound with [var].
+      Sequence of searches is the following:
+      1) among write events that occured at current step in the same context,
+         with the same variable; we don't need to emit event in this case
+      2) among read events that occured at current step in the same context,
+         with the same variable; will emit register read in this case
+      3) among read events that occures at current step, in other context,
+         with the same variable; will emit register read in this case *)
+    method private find_var_data var =
       SM.get () >>= fun ctxt ->
-      match find_reg_read (self_events ctxt) (same_var var) with
-      | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-      | None ->
-        match find_reg_read (other_events ctxt) (same_var var) with
-        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-        | None -> super#lookup var
-
-    (** [lookup var] - returns a result, bound with variable.
-        Search starts from self events, if it was write access to given
-        variable at current step. And if it was, then result of write
-        access returned and no read event is emitted.
-        Otherwise searching continues as written above for [resolve_var],
-        with emitting register_read event. *)
+      let searches = [
+        (fun () -> find_reg_write (self_events ctxt)  (same_var var)), false;
+        (fun () -> find_reg_read  (self_events ctxt)  (same_var var)), true;
+        (fun () -> find_reg_read  (other_events ctxt) (same_var var)), true;
+      ] in
+      SM.return (find_move_data searches)
+
     method! lookup var : 'a r =
       SM.get () >>= fun ctxt ->
-      match find_reg_write (self_events ctxt) (same_var var) with
-      | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-      | None ->
-        self#resolve_var var >>= fun r ->
-        match value r with
-        | Bil.Imm data ->
-          if not (Var.is_virtual var) then
-            self#update_event (create_reg_read var data) >>= fun () ->
-            SM.return r
-          else SM.return r
-        | Bil.Mem _ | Bil.Bot -> SM.return r
+      self#find_var_data var >>= function
+      | None -> super#lookup var
+      | Some (data, need_emit) ->
+        self#eval_exp (Bil.int data) >>= fun r ->
```

Comment:
It's not clear to me, why are we hiding (not emitting) an event of a variable read, if we obtained its value by looking up in the write event of the trace. This was still an access to a variable, and it doesn't really matter, what was the source of the variable value.

Reply:
imagine we have an instruction with the next write event: `RAX <= 42`, and then this value is used to a calculation of flags. So there would not be a `read` event from register, but just a using of its value

Reply:
The code models a behavior that is more general, that your example describes. The code basically says:

"if there is a read from X, then lookup its value in the trace. If we found in a trice a write to X then do not emit the read event, otherwise do emit"

I do not believe, that this general rule is correct, even if it solves your particular case. It looks like, that you're trying to pick into the trace, to grab the semantics of current instructions, that is not right, it is the same as trying to pry into the set of answers. The set of events emitted by the BIL interpreter should not depend on the set of events of the trace that is used for verification. 

Given your particular example, I don't see any problems with emitting the read event. Moreover, I think it is wrong not to emit the event. If a flag calculation depends on the value of the RAX register, then the RAX register must be read. Also, there is no difference between using a register and reading from a register. They are synonyms. Every time a value of a register is obtained, a read event should be emitted.